l1-contracts/contracts/bridge/asset-tracker/L1AssetTracker.sol

```diff
@@ -0,0 +1,369 @@
+            .decodeTokenBalanceMigrationData(_finalizeWithdrawalParams.message);
+        require(
+            functionSignature == IAssetTrackerDataEncoding.receiveMigrationOnL1.selector,
+            InvalidFunctionSignature(functionSignature)
+        );
+
+        require(assetMigrationNumber[data.chainId][data.assetId] < data.migrationNumber, InvalidAssetId(data.assetId));
+
+        uint256 currentSettlementLayer = _bridgehub().settlementLayer(data.chainId);
+        uint256 fromChainId;
+        uint256 toChainId;
+
+        if (data.isL1ToGateway) {
+            uint256 chainMigrationNumber = _getChainMigrationNumber(data.chainId);
+            /// We check the chainMigrationNumber to make sure the message is from a previous token migration.
+            require(
+                chainMigrationNumber == data.migrationNumber,
+                InvalidChainMigrationNumber(chainMigrationNumber, data.migrationNumber)
+            );
+            /// In this case the TokenBalanceMigrationData data might be malicious.
+            /// We check the chainId to match the finalizeWithdrawalParams.chainId.
+            /// We check the assetId, tokenOriginChainId, originToken with an assetIdCheck.
+            /// The amount might be malicious, but that poses a restriction on users of the chain, not other chains.
+            /// The AssetTracker cannot protect individual users only other chains. Individual users rely on the proof system.
+            /// The last field is migrationNumber, which cannot be abused.
+
+            require(currentSettlementLayer != block.chainid, NotMigratedChain());
+            require(data.chainId == _finalizeWithdrawalParams.chainId, InvalidWithdrawalChainId());
+
+            // we check parity here to make sure that we migrated the token balance back to L1 from Gateway.
+            // this is needed to ensure that the chainBalance on the Gateway AssetTracker is currently 0.
+            // In the future we might initialize chains on GW. So we subtract from chainMigrationNumber.
+            require(
+                (chainMigrationNumber - assetMigrationNumber[data.chainId][data.assetId]) % 2 == 1,
+                InvalidMigrationNumber(chainMigrationNumber, assetMigrationNumber[data.chainId][data.assetId])
+            );
+
+            /// We check the assetId to make sure the chain is not lying about it.
+            DataEncoding.assetIdCheck(data.tokenOriginChainId, data.assetId, data.originToken);
+
+            fromChainId = data.chainId;
+            toChainId = currentSettlementLayer;
+        } else {
+            /// In this case we trust the TokenBalanceMigrationData data and the settlement layer = Gateway to be honest.
+            /// If the settlement layer is compromised, other chains settling on L1 are not compromised, only chains settling on Gateway.
+
+            require(currentSettlementLayer == block.chainid, NotMigratedChain());
```

Comment:
if a chain migrates to SL_A, then migrates to L1, then migrates to SL_B, 

then it wont be able to migrate the token from SL_A, to L1 to migrate it further to SL_B.

Reply:
My assumption here was kind of that the chain admin will migrate the important tokens, and the rest will be left on the GW, but it is not ideal, I agree. 

The best solution would be to whitelist specific GW's per chain, i.e. GWs that it trusts.